### PR TITLE
feat: add pprof register and config option for chaos dashboard.

### LINKED ISSUE #1053: Enable pprof in Chaos Dashboard component
## Feature Request

**Is your feature request related to a problem? Please describe:**

Sometimes we need to locate some Chaos Mesh issues, such as performance, memory leaks, etc. I think pprof can help us. In addition, we have enabled pprof in other components except chaos-dashboard, so I also need to enable pprof in the chaos-dashboard component.

### PR DESCRIPTION
**Pull request by Colstuwjx**
Signed-off-by: colstuwjx <colstuwjx@gmail.com>

### What problem does this PR solve?

Fixed #1053 .

### What is changed and how does it work?
Add an env config option `ENABLE_PPROF` for controlling whether enable the chaos-dashboard pprof.

### Checklist

Tests

- [ ] Unit test
- [ ] E2E test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Breaking backward compatibility

Related changes

- [x] Need to update the documentation

### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by Colstuwjx** [comment]
I'm trying to test this feature on my laptop, but while I have been setup the chaos mesh via `./install.sh --runtime containerd` according the [CONTRIBUTING.md](https://github.com/chaos-mesh/chaos-mesh/blob/master/CONTRIBUTING.md#step-3-start-chaos-mesh-locally-and-do-manual-tests), but it seems that chaos-dashboard was not running healthy, see the error logs below:

```
2020-10-17T11:09:35.206Z	ERROR	experimentStore	failed to delete all incomplete experiments	{"error": "no such table: events"}
github.com/go-logr/zapr.(*zapLogger).Error
	/root/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
github.com/chaos-mesh/chaos-mesh/pkg/store/experiment.NewStore
	/src/pkg/store/experiment/experiment.go:38
reflect.Value.call
	/usr/local/go/src/reflect/value.go:460
reflect.Value.Call
	/usr/local/go/src/reflect/value.go:321
go.uber.org/dig.(*node).Call
	/root/go/pkg/mod/go.uber.org/dig@v1.9.0/dig.go:667

go.uber.org/dig.paramSingle.Build
	/root/go/pkg/mod/go.uber.org/dig@v1.9.0/param.go:245
(/src/pkg/store/experiment/experiment.go:51)
go.uber.org/dig.paramList.BuildList
[2020-10-17 11:09:35]  no such table: events
	/root/go/pkg/mod/go.uber.org/dig@v1.9.0/param.go:201
go.uber.org/dig.(*node).Call
	/root/go/pkg/mod/go.uber.org/dig@v1.9.0/dig.go:658
go.uber.org/dig.paramSingle.Build
	/root/go/pkg/mod/go.uber.org/dig@v1.9.0/param.go:245
go.uber.org/dig.paramList.BuildList
	/root/go/pkg/mod/go.uber.org/dig@v1.9.0/param.go:201
go.uber.org/dig.(*node).Call
	/root/go/pkg/mod/go.uber.org/dig@v1.9.0/dig.go:658
go.uber.org/dig.paramSingle.Build
	/root/go/pkg/mod/go.uber.org/dig@v1.9.0/param.go:245
go.uber.org/dig.paramList.BuildList
	/root/go/pkg/mod/go.uber.org/dig@v1.9.0/param.go:201
go.uber.org/dig.(*Container).Invoke
	/root/go/pkg/mod/go.uber.org/dig@v1.9.0/dig.go:389
go.uber.org/fx.(*App).executeInvokes
	/root/go/pkg/mod/go.uber.org/fx@v1.12.0/app.go:659
go.uber.org/fx.New
	/root/go/pkg/mod/go.uber.org/fx@v1.12.0/app.go:438
main.main
	/src/cmd/chaos-dashboard/main.go:80
runtime.main
	/usr/local/go/src/runtime/proc.go:203
2020-10-17T11:09:35.548Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": ":8080"}
2020/10/17 11:09:35 [Fx] INVOKE		github.com/chaos-mesh/chaos-mesh/pkg/apiserver/experiment.Register()
2020/10/17 11:09:35 [Fx] INVOKE		github.com/chaos-mesh/chaos-mesh/pkg/apiserver/event.Register()
[GIN-debug] POST   /api/common/pods          --> github.com/chaos-mesh/chaos-mesh/pkg/apiserver/common.(*Service).listPods-fm (4 handlers)
[GIN-debug] GET    /api/common/namespaces    --> github.com/chaos-mesh/chaos-mesh/pkg/apiserver/common.(*Service).getNamespaces-fm (4 handlers)
```

It will be appreciated if someone could help me clear this error and confirm whether the pprof enabled and works...
Thanks.

**Comment by fewdan** [comment]
@Colstuwjx
Move "pprof.Register(r)" from “serverRegister” to “newEngine” , like below:
image
image

It seems to move the position of "pprof.Register(r)" and it can run normally.

File: pkg/apiserver/server.go [hunk 1]
```diff
@@ -49,6 +50,11 @@ var (
 func serverRegister(r *gin.Engine, conf *config.ChaosDashboardConfig) {
 	listenAddr := fmt.Sprintf("%s:%d", conf.ListenHost, conf.ListenPort)
 
+	if conf.EnablePprof {
```
Comment (fewdan) [inline pkg/apiserver/server.go:53]: As I commented above, maybe we need to move the position of "pprof.Register(r)".

**Review by fewdan — CHANGES_REQUESTED** [review]

File: pkg/apiserver/server.go (hunk 1 above)
Comment (Colstuwjx) [reply pkg/apiserver/server.go:53]: At the first time, I did this, but I found that we couldn't consume the `conf.EnablePprof` such flag anymore, right?

Comment (fewdan) [reply pkg/apiserver/server.go:53]: Oh oh, let me take a look.

Comment (fewdan) [reply pkg/apiserver/server.go:53]: Maybe you can do something like below :
`
func newEngine (conf *config.ChaosDashboardConfig) *gin.Engine {
	r := gin.Default()

	if conf.EnablePprof {
		// default is "debug/pprof"
		pprof.Register(r)
	}
        ....................
}
`

**Review by yeya24 — COMMENTED** [review]
Just curious why not remove the flag and make pprof enabled as default?

**Comment by fewdan** [comment]
> Just curious why not remove the flag and make pprof enabled as default?

@yeya24 
I think that since pprof is introduced, a flag is needed to control it.

**Comment by yeya24** [comment]
> > Just curious why not remove the flag and make pprof enabled as default?
> 
> @yeya24
> I think that since pprof is introduced, a flag is needed to control it.

Can you please explain why? A pprof HTTP endpoint doesn't have any performance penalty if we don't call it. If you check other applications with pprof, it is enabled by default and there is no flag to control this.

**Comment by fewdan** [comment]
> Can you please explain why? A pprof HTTP endpoint doesn't have any performance penalty if we don't call it. If you check other applications with pprof, it is enabled by default and there is no flag to control this.

My concern before is that it will affect performance, if there is no performance loss, it will be enabled by default.

**Comment by Colstuwjx** [comment]
> Can you please explain why? A pprof HTTP endpoint doesn't have any performance penalty if we don't call it. If you check other applications with pprof, it is enabled by default and there is no flag to control this.

OK, now I just register the pprof interface without adding the flag, see the capture for demo:

image

image

The url is: http://127.0.0.1:2333/debug/pprof/ .

**Review by yeya24 — APPROVED** [review]
LGTM

**Review by fewdan — APPROVED** [review]
LGTM

**Pull request merged**